From a06c6885a1a7483f61c904089013ddfabbcf9db1 Mon Sep 17 00:00:00 2001
From: Nicolas Le Bayon <nicolas.le.bayon@foss.st.com>
Date: Tue, 26 Nov 2024 18:28:59 +0100
Subject: [PATCH 744/841] refactor(st-crypto): split RNG NIST configuration

Separate each RNG NIST configuration bit field and add a macro to
compute global value.
Remove deprecated definitions.
Use BIT_32 and GENMASK_32 macros, as they're adapted to 32 bit registers.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git c8428d9eec3eef80b0bdb12861f18a62e78367c7 refactor(st-crypto): split RNG NIST configuration]
Signed-off-by: Nicolas Le Bayon <nicolas.le.bayon@foss.st.com>
Change-Id: I9db5f874c7e33cd41dc16b755b3cfdc5053fc995
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/421390
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
Tested-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
---
 drivers/st/crypto/stm32_rng.c | 64 +++++++++++++++++++----------------
 1 file changed, 35 insertions(+), 29 deletions(-)

diff --git a/drivers/st/crypto/stm32_rng.c b/drivers/st/crypto/stm32_rng.c
index be2cb5c54..04bf7e6f2 100644
--- a/drivers/st/crypto/stm32_rng.c
+++ b/drivers/st/crypto/stm32_rng.c
@@ -20,50 +20,55 @@
 #include <platform_def.h>
 
 #if STM32_RNG_VER == 2
-#define DT_RNG_COMPAT		"st,stm32-rng"
+#define DT_RNG_COMPAT			"st,stm32-rng"
 #endif
 #if STM32_RNG_VER == 4
-#define DT_RNG_COMPAT		"st,stm32mp13-rng"
+#define DT_RNG_COMPAT			"st,stm32mp13-rng"
 #endif
-#define RNG_CR			0x00U
-#define RNG_SR			0x04U
-#define RNG_DR			0x08U
+#define RNG_CR				0x00U
+#define RNG_SR				0x04U
+#define RNG_DR				0x08U
 #if STM32_RNG_VER == 4
-#define RNG_HTCR		0x10U
+#define RNG_HTCR			0x10U
 #endif
 
-#define RNG_CR_RNGEN		BIT(2)
-#define RNG_CR_IE		BIT(3)
-#define RNG_CR_CED		BIT(5)
-#define RNG_CR_CLKDIV		GENMASK(19, 16)
-#define RNG_CR_CLKDIV_SHIFT	16U
-#define RNG_CR_CONDRST		BIT(30)
-
-#define RNG_SR_DRDY		BIT(0)
-#define RNG_SR_CECS		BIT(1)
-#define RNG_SR_SECS		BIT(2)
-#define RNG_SR_CEIS		BIT(5)
-#define RNG_SR_SEIS		BIT(6)
+#define RNG_CR_RNGEN			BIT_32(2)
+#define RNG_CR_CED			BIT_32(5)
+#if STM32_RNG_VER == 4
+#define RNG_CR_RNG_CONFIG3_SHIFT	8U
+#define RNG_CR_RNG_CONFIG2_SHIFT	13U
+#define RNG_CR_CLKDIV			GENMASK_32(19, 16)
+#define RNG_CR_CLKDIV_SHIFT		16U
+#define RNG_CR_RNG_CONFIG1_SHIFT	20U
+#define RNG_CR_CONDRST			BIT_32(30)
+#endif
 
-#define RNG_TIMEOUT_US		100000U
-#define RNG_TIMEOUT_STEP_US	10U
+#define RNG_SR_DRDY			BIT_32(0)
+#define RNG_SR_SECS			BIT_32(2)
+#define RNG_SR_SEIS			BIT_32(6)
 
-#define TIMEOUT_US_1MS		1000U
+#define RNG_TIMEOUT_US			100000U
+#define RNG_TIMEOUT_STEP_US		10U
 
-#define RNG_NIST_CONFIG_A	0x00F40F00U
-#define RNG_NIST_CONFIG_B	0x01801000U
-#define RNG_NIST_CONFIG_C	0x00F00D00U
-#define RNG_NIST_CONFIG_MASK	GENMASK(25, 8)
+#define TIMEOUT_US_1MS			1000U
 
 #if STM32_RNG_VER == 4
+#define RNG_NIST_CONFIG1		0xFU
+#define RNG_NIST_CONFIG2		0x0U
+#define RNG_NIST_CONFIG3		0xFU
+#define RNG_NIST_CONFIG(x,y,z)		(((x) << RNG_CR_RNG_CONFIG1_SHIFT) | \
+					 ((y) << RNG_CR_RNG_CONFIG2_SHIFT) | \
+					 ((z) << RNG_CR_RNG_CONFIG3_SHIFT))
+#define RNG_NIST_CONFIG_MASK		GENMASK_32(25, 8)
+
 #if STM32_RNG_VER_MINOR == 2
-#define RNG_HTCFG_CONFIG	0x000072ACU /* Reset value */
+#define RNG_HTCFG_CONFIG		0x000072ACU /* Reset value */
 #else
-#define RNG_HTCFG_CONFIG	0x0000AAC7U
+#define RNG_HTCFG_CONFIG		0x0000AAC7U
 #endif
 #endif
 
-#define RNG_MAX_NOISE_CLK_FREQ	48000000U
+#define RNG_MAX_NOISE_CLK_FREQ		48000000U
 
 struct stm32_rng_instance {
 	uintptr_t base;
@@ -131,7 +136,8 @@ static int stm32_rng_enable(void)
 
 	/* Update configuration fields */
 	mmio_clrsetbits_32(stm32_rng.base + RNG_CR, RNG_NIST_CONFIG_MASK,
-			   RNG_NIST_CONFIG_A | RNG_CR_CONDRST | RNG_CR_CED);
+			   RNG_NIST_CONFIG(RNG_NIST_CONFIG1, RNG_NIST_CONFIG2, RNG_NIST_CONFIG3) |
+			   RNG_CR_CONDRST | RNG_CR_CED);
 
 	mmio_clrsetbits_32(stm32_rng.base + RNG_CR, RNG_CR_CLKDIV,
 			   (clock_div << RNG_CR_CLKDIV_SHIFT));
-- 
2.47.2

